runtime/metrics/src/nostd.rs

```diff
@@ -0,0 +1,112 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+const TRACING_TARGET: &'static str = "metrics";
+
+use parity_scale_codec::Encode;
+use primitives::v1::{
+	RuntimeMetricLabelValues, RuntimeMetricOp, RuntimeMetricRegisterParams, RuntimeMetricUpdate,
+};
+
+pub struct CounterVec {
+	name: &'static str,
+	label_values: Option<RuntimeMetricLabelValues>,
+}
+
+pub struct Counter {
+	name: &'static str,
+}
+
+fn emit_metric(metric_op: &[u8]) {
+	// `from_utf8_unchecked` is safe in this context:
+	// We cannot guarantee that the input slice is a valid UTF-8, but the `bs58` encoding
+	// provides that guarantee forward. We just need to `transmute` as the layout of `&str`
+	// and `&[u8]` is the same.
+
+	unsafe {
+		let metric_op = bs58::encode(sp_std::str::from_utf8_unchecked(&metric_op)).into_string();
```

Comment:
the only unsafe function here is `sp_std::str::from_utf8_unchecked(&metric_op)`, where `metric_op` is input `&[u8]`, so I don't get the safety proof and the reason to use this function in the first place? premature optimization?

Reply:
Imho we should still use a bytestring / byteslice for the key and avoid all of the string based stuff in the runtime

Reply:
This is not an optimization in any case, it's just making sure our encoded metric updates comply with the tracing interfaces - the tracing API which requires that fields of Events/Spans are all Strings or integers. It doesn't allow any other types, such as the ones we are interested in.

I don't think we can avoid this, unless we make changes to the tracing API.  

In our specific context, all of the `Event` parameters are further formatted into a single string and handled in the `TraceHandler` impl (they are string values).

Does this make sense ?

Reply:
`bs58::encode` accepts `AsRef<[u8]>`: https://docs.rs/bs58/0.4.0/bs58/fn.encode.html
why do we cast it to `&str` then?

Reply:
In that case, you are correct, and we can't.

Reply:
Yeah, got it, it makes sense.

Reply:
I missed the fact that it can bs58 take u8 slice as input.